flow.py

```diff
@@ -57,17 +62,70 @@ def dest(self):
     def dest(self, dest):
         raise AttributError("Cannot modify a flow's destination node.")
 
-    def make_packet(self, id, src, dest, data):
+    def check_last(self):
+        """Method that checks if all packets have been acknowledged and
+        updating the NetworkSimulator object accordingly
+        """
+        if self.last_acknowledged == self.num_packets:
+            self.ns.num_active_flows -= 1
+
+    def update_flow(a_packet):
+        """Upon receiving an acknowledgement packet, updating the flow's
+        attributes
+
+        Args:
+            a_packet (AcknowledgementPacket): Packet being sent
+                back from host
+        """
+        print("Flow: updated with acknowledgement packet")
+        self.last_acknowledged += 1
+        self.check_last()
+        self.unacknowledged_packets.remove(a_packet)
+
+
+    def make_data_packet(self, packet_id, src, dest, packet_size, data):
         """Method makes the packet and triggers the send_packet method for the
         host if applicable
 
         Args:
-            id (int): The unique id identifying the packet
+            packet_id (string): Unique id identifying the packet
             src (Node): The packet's source node
             dest (Node): The packet's destination node
+            packet_size (float): The packet's size in bits
+            flow_id (string): Unique id indicating packet
             data (any datatype): The data in the packet
         """
-        pass
+        print("Flow: made data packet")
+        new_packet = DataPacket(ns, packet_id, src, dest, self.flow_id,
+            packet_size, data)
+        ns.flows[new_packet.dest].send_packet(new_packet)
+
+    def make_acknowledgment_packet(self, packet_id, src, dest, packet_size, acknowledge_id):
+        """Method makes the packet and triggers the send_packet method for the
+        host if applicable
+
+        Args:
+            packet_id (string): Unique id identifying the packet
+            src (Node): The packet's source node
+            dest (Node): The packet's destination node
+            packet_size (float): The packet's size in bits
+            flow_id (string): Unique id indicating packet
+            acknowledge_id (string): Unique id identifying packet being acknowledged
+        """
+        print("Flow: made acknowledgement packet")
+        new_packet = AcknowledgementPacket(ns, packet_id, src, dest,
+            packet_size, self.flow_id, acknowledge_id)
+        ns.flows[new_packet.dest].send_packet(new_packet)
+
+    def receive_packet(self, packet):
+        """Method receives a given packet: acknowledges packet if it's a
+        data packet, updates the flow if it's an acknowledgement packet
+        """
+        print("Flow: received packet")
+        if isinstance(packet, DataPacket):
```

Comment:
If it's a data packet, you send an acknowledgement packet.

if it's an ack, you update_flow